drivers/audio/wm8904.c

```diff
@@ -0,0 +1,759 @@
+
+	/* WSEQ_ABORT=0, WSEQ_START=1, WSEQ_START_INDEX=00_0000 */
+	wm8904_write_reg(dev, WM8904_WRT_SEQUENCER_3, 0x0100);
+
+	do {
+		wm8904_read_reg(dev, WM8904_WRT_SEQUENCER_4, &value);
+	} while (((value & 1U) != 0U));
+
+	/* TOCLK_RATE_DIV16=0, TOCLK_RATE_x4=1, SR_MODE=0, MCLK_DIV=1
+	 * (Required for MMCs: SGY, KRT see erratum CE000546)
+	 */
+	wm8904_write_reg(dev, WM8904_CLK_RATES_0, 0xA45F);
+
+	/* INL_ENA=1, INR ENA=1 */
+	wm8904_write_reg(dev, WM8904_POWER_MGMT_0, 0x0003);
+
+	/* HPL_PGA_ENA=1, HPR_PGA_ENA=1 */
+	wm8904_write_reg(dev, WM8904_POWER_MGMT_2, 0x0003);
+
+	/* DACL_ENA=1, DACR_ENA=1, ADCL_ENA=1, ADCR_ENA=1 */
+	wm8904_write_reg(dev, WM8904_POWER_MGMT_6, 0x000F);
+
+	/* ADC_OSR128=1 */
+	wm8904_write_reg(dev, WM8904_ANALOG_ADC_0, 0x0001);
+
+	/* DACL_DATINV=0, DACR_DATINV=0, DAC_BOOST=00, LOOPBACK=0, AIFADCL_SRC=0,
+	 * AIFADCR_SRC=1, AIFDACL_SRC=0, AIFDACR_SRC=1, ADC_COMP=0, ADC_COMPMODE=0,
+	 * DAC_COMP=0, DAC_COMPMODE=0
+	 */
+	wm8904_write_reg(dev, WM8904_AUDIO_IF_0, 0x0050);
+
+	/* DAC_MONO=0, DAC_SB_FILT-0, DAC_MUTERATE=0, DAC_UNMUTE RAMP=0,
+	 * DAC_OSR128=1, DAC_MUTE=0, DEEMPH=0 (none)
+	 */
+	wm8904_write_reg(dev, WM8904_DAC_DIG_1, 0x0040);
+
+	/* Enable DC servos for headphone out */
+	wm8904_write_reg(dev, WM8904_DC_SERVO_0, 0x0003);
+
+	/* HPL_RMV_SHORT=1, HPL_ENA_OUTP=1, HPL_ENA_DLY=1, HPL_ENA=1,
+	 * HPR_RMV_SHORT=1, HPR_ENA_OUTP=1, HPR_ENA_DLY=1, HPR_ENA=1
+	 */
+	wm8904_write_reg(dev, WM8904_ANALOG_HP_0, 0x00FF);
+
+	/* CP_DYN_PWR=1 */
+	wm8904_write_reg(dev, WM8904_CLS_W_0, 0x0001);
+
+	/* CP_ENA=1 */
+	wm8904_write_reg(dev, WM8904_CHRG_PUMP_0, 0x0001);
+
+	wm8904_protocol_config(dev, cfg->dai_type);
+	wm8904_update_reg(dev, WM8904_CLK_RATES_2, (uint16_t)(1UL << 14U),
+			  (uint16_t)(dev_cfg->clock_source));
+
+	if (dev_cfg->clock_source == 0) {
+		int err = clock_control_on(dev_cfg->mclk_dev, dev_cfg->mclk_name);
+
+		if (err < 0) {
+			LOG_ERR("MCLK clock source enable fail: %d", err);
+		}
+
+		err = clock_control_get_rate(dev_cfg->mclk_dev, dev_cfg->mclk_name,
+					     &cfg->mclk_freq);
+		if (err < 0) {
+			LOG_ERR("MCLK clock source freq acquire fail: %d", err);
+		}
+	}
+
+	wm8904_audio_fmt_config(dev, &cfg->dai_cfg, cfg->mclk_freq);
+
+	if ((cfg->dai_cfg.i2s.options & I2S_OPT_FRAME_CLK_MASTER) == I2S_OPT_FRAME_CLK_MASTER) {
+		wm8904_set_master_clock(dev, &cfg->dai_cfg, cfg->mclk_freq);
+	} else {
+		/* BCLK/LRCLK default direction input */
+		wm8904_update_reg(dev, WM8904_AUDIO_IF_1, 1U << 6U, 0U);
+		wm8904_update_reg(dev, WM8904_AUDIO_IF_3, (uint16_t)(1UL << 11U), 0U);
+	}
+
+	switch (cfg->dai_route) {
+	case AUDIO_ROUTE_PLAYBACK:
+		wm8904_configure_output(dev);
+		break;
+
+	case AUDIO_ROUTE_CAPTURE:
+		wm8904_configure_input(dev);
+		break;
+
+	case AUDIO_ROUTE_PLAYBACK_CAPTURE:
+		wm8904_configure_output(dev);
+		wm8904_configure_input(dev);
+		break;
+
+	default:
+		break;
+	}
+
+	return 0;
+}
+
+static void wm8904_start_output(const struct device *dev)
+{
```

Comment:
Why are these 2 functions empty? 
No logging info, nothing?

Reply:
@iuliana-prodan These functions correspond to the `audio_codec_start_output` and `audio_codec_stop_output` syscalls in `include/zephyr/audio/codec.h`. These syscall functions don't permit NULL values in the API structure, thus these functions are mandatory to implement. My goal was to change the `codec` API as little as possible, so that's why I left it as it is.

Reply:
ok.